### ISSUE TITLE
YOLOV5 speed 

### ISSUE THREAD

**Issue by mfaizan-ai** [ultralytics/yolov5#7975]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi Glenn, I hope you are doing well. I am using YOLOv5m, and I scaled it with model depth multiple = 0.50, and model channel multiple = 0.625, which is the average of yolov5s and yolov5m constants. I named this yolov5c (custom). I have trained yolov5m, yolov5s, and yolov5c for performance comparison. I want to improve the speed of yolov5m. After running the above three models for 300 epochs on 3 classes with 8k images on the test set, I got the following results. I used default hyperparameters. I have 2000 images for evaluation, 1000 for the test, and 100 val.

Following is the models comparison table
![yolov5c_speed_question](https://user-images.githubusercontent.com/61932757/170244841-697eb72a-61c5-4fa9-b63c-073d383337b1.png)

and improvement speed improvement table
![improvemet](https://user-images.githubusercontent.com/61932757/170244904-0af48644-1551-4052-a9aa-97b38f206a26.png)

My question is when I run the yolov5 val.py script on --the task test, the speed differs on the same folder with identical settings, why is that? each time running inferencing on the test folder results in different speed values, which are different from what I get from the detect.py script. 

and I have written my own script to measure speed on 416x416 resolution image. I ran this 10000 times and averaged the result, which is different from the yolov5 evaluation script speed measurement.

### Additional

The code for my speed measurement script is given here. I ran it on a single image with 416x416 resolution.

import argparse
from os 
import device_encoding
import time
import numpy as np
import cv2
import argparse
import torch

def parsing_args():
    '''parsing args'''
    parser = argparse.ArgumentParser()
    parser.add_argument('-t', '--times', type = int, help= 'number of times to run an image for inference')
    parser.add_argument('--image', help="path to an image for inference")
    parser.add_argument('--weights', help='path to weights')
    parser.add_argument('--imgsz', type = int, help= 'image size for resizing')
    parser.add_argument('--resize', action='store_false',  help='resize the image into required shape')
    parser.add_argument('--img_name', type=str, help="name of the image")
    parser.add_argument('--device',  default= "", help='cuda device, i.e. 0 or 0,1,2,3 or cpu')
    return parser.parse_args()

def inference(num_times, image, weights, device):
    '''run inference on an image and calculate it's speed'''
    # device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    model = torch.hub.load('ultralytics/yolov5', 'custom', path= weights)
    img = cv2.imread(image, cv2.IMREAD_UNCHANGED)
    img_dash = torch.from_numpy(img)
    img_dash = img_dash[None, :]
    img_dash = img_dash.permute(0, 3, 1, 2) #X_train = X_train.permute(2, 0, 1)
    img_dash = img_dash.to(device)

   
    #warmup
    for i in range(100):

        _ = model(img_dash)
    
    #inference
    time_record = 0
    for i in range(num_times):
        tic = time.time()   #time in ms
        results = model(img_dash)
        toc = time.time()
        time_duration = toc- tic
        time_record+=time_duration
        print(f'{time_duration*1: 0.3f} ms')
    average_time = (time_record)/num_times
    print(f'shape of an image: {img.shape}')
    return average_time

def resize(image, dim, img_name):
    '''resize into 640x640'''
    img = cv2.imread(image, cv2.IMREAD_UNCHANGED)
    resized_img = cv2.resize(img, (dim, dim), interpolation=cv2.INTER_AREA)
    cv2.imwrite(f'{img_name}', resized_img)

def main():
    args = parsing_args()
    if args.weights:
        avg_time = inference(args.times, args.image, args.weights,args.device)
    # if args.resize:
    #     resize(args.image, args.imgsz, args.img_name)
    #     print('image has been saved!!')
    print(f'average time taken by the model for running an image {args.times: .3f} times: {avg_time*1000} ms')
    print('finsined!!')

if __name__ == '__main__':
    main()

**Comment by glenn-jocher** [comment]
@faizan1234567 commands to reproduce official speeds are indicated below README table here:

<img width="1414" alt="Screen Shot 2022-05-25 at 6 18 36 PM" src="https://user-images.githubusercontent.com/26833433/170310700-fd464f31-4bab-45fd-b368-a3c38446543e.png">

**Comment by mfaizan-ai** [comment]
is this speed on the whole dataset? I guess it's on the evaluation set, but I don't know if it's on the validation set or the test set?

**Comment by glenn-jocher** [comment]
val

**Comment by mfaizan-ai** [comment]
Is my code for speed measurement on a single image fine? I have put both model and image to the device i.e. Cuda. Glenn, is there any way to measure the speed of each layer during inference for experiment purposes?

**Comment by glenn-jocher** [comment]
@faizan1234567 yes you can profile layers with 
```
python models/yolo.py --cfg yolov5s.yaml --line-profile
```
